core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,481 @@
+from sklearn.svm import SVC
+
+# pylint: disable=invalid-name
+def _get_tokens(program):
+    """Generate tokens for program using tokenize module."""
+    for tid, tname, _, _, _ in tokenize.generate_tokens(
+            StringIO(program).readline):
+        yield (tid, tname)
+
+
+def _cv_tokenizer(program):
+    """Tokenize Python program for CountVectorizer."""
+    token_program = []
+    for tid, tname in _get_tokens(program):
+        if tid == token.N_TOKENS or tid == 54:
+            continue
+        elif tid == token.NAME:
+            if tname in keyword.kwlist:
+                token_program.append(tname)
+            else:
+                token_program.append('V')
+        else:
+            token_program.append(tname)
+
+    return token_program
+
+
+def _generate_token_to_id(data, threshold=5):
+    """Generates a list of valid tokens and assigns a unique ID to each
+    token."""
+    # All unique tokens and number of time they occur in dataset. A token
+    # can be a keyword, an identifier, a constant, an operator.
+    alphabet = {}
+
+    for pid in data:
+        program = data[pid]['source']
+        for tid, tname in _get_tokens(program):
+            # If tid is tokens.NAME then only add if it is a python keyword.
+            # Treat all variables and methods same.
+            if tid == token.N_TOKENS or tid == 54:
+                continue
+            elif tid == token.NAME:
+                if tname in keyword.kwlist:
+                    alphabet[tname] = alphabet.get(tname, 0) + 1
+            else:
+                alphabet[tname] = alphabet.get(tname, 0) + 1
+
+    # Consider only those tokens which occur for more than threshold times
+    # in entire dataset.
+    valid_tokens = [k for k, v in alphabet.iteritems() if v > threshold]
+    token_to_id = dict(zip(valid_tokens, xrange(0, len(valid_tokens))))
+
+    # Add 'UNK' in token_to_id. This will be used to replace any token
+    # occurring in program which is not in valid_token.
+    token_to_id['UNK'] = len(token_to_id)
+
+    # Add 'V' in token_to_id. This token will be used to replace all
+    # variables and methods in program.
+    token_to_id['V'] = len(token_to_id)
+    return token_to_id
+
+
+def _tokenize_data(data):
+    """Tokenize Python programs in dataset for winnowing."""
+    token_to_id = _generate_token_to_id(data)
+
+    # Tokenize all programs in dataset.
+    for program_id in data:
+        program = data[program_id]['source']
+        token_program = []
+        for tid, tname in _get_tokens(program):
+            if tid == token.N_TOKENS or tid == 54:
+                continue
+            elif tid == token.NAME:
+                if tname in keyword.kwlist:
+                    if tname not in token_to_id:
+                        token_program.append('UNK')
+                    else:
+                        token_program.append(tname)
+                else:
+                    token_program.append('V')
```

Comment:
Especially more useful to assign 'V' to a variable if you have this string multiple times through the code. same for 'UNK'.